src/net/atlas/onchain.rs

```diff
@@ -0,0 +1,235 @@
+use super::AtlasDB;
+
+use chainstate::burn::db::sortdb::SortitionDB;
+use chainstate::stacks::boot;
+use chainstate::stacks::db::StacksChainState;
+use chainstate::stacks::StacksBlockId;
+use net::connection::ConnectionOptions;
+use net::Error as net_error;
+use rusqlite::Row;
+use util::db::Error as db_error;
+use util::hash::Hash160;
+use util::hash::MerkleHashFunc;
+use vm::{
+    clarity::{ClarityConnection, ClarityReadOnlyConnection},
+    costs::{ExecutionCost, LimitedCostTracker},
+    database::{ClarityDatabase, ClaritySerializable, MarfedKV, STXBalance},
+    types::{
+        PrincipalData, QualifiedContractIdentifier, SequenceData, StandardPrincipalData, TupleData,
+    },
+    ClarityName, ContractName, Value,
+};
+
+use std::collections::HashMap;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OnchainAttachmentsInventory {
+    pub tip: StacksBlockId,
+    pub info: OnchainAttachmentsInventoryInfo,
+    pub pages_indexes: Vec<u32>,
+    pub pages: HashMap<u32, OnchainAttachmentPage>,
+}
+
+impl OnchainAttachmentsInventory {
+    pub fn empty() -> OnchainAttachmentsInventory {
+        OnchainAttachmentsInventory {
+            tip: StacksBlockId([0x00; 32]),
+            info: OnchainAttachmentsInventoryInfo::empty(),
+            pages_indexes: vec![],
+            pages: HashMap::new(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct OnchainAttachmentsInventoryInfo {
+    pub pages_count: u32,
+    pub last_page_len: u32,
+    pub page_size: u32,
+}
+
+impl OnchainAttachmentsInventoryInfo {
+    pub fn empty() -> OnchainAttachmentsInventoryInfo {
+        OnchainAttachmentsInventoryInfo {
+            pages_count: 0,
+            last_page_len: 0,
+            page_size: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct OnchainAttachmentPage {
+    pub index: u32,
+    pub entries: Vec<Hash160>,
+}
+
+pub struct OnchainInventoryLookup {}
+
+impl OnchainInventoryLookup {
```

Comment:
Just confirming my understanding -- the reason we want to store content hashes this way is because attachments must be somewhat stacks fork-aware, right?

Reply:
Yes, it's a pagination mechanism that helps nodes sharing attachments inventories, for-aware